draft-pda-protocol.md

```diff
@@ -573,116 +609,281 @@ PA protocol. The structure of the leader share is similarly protocol specific:
+the following algorithm:
+
+~~~
+state = CreateState(param)
+while not state.Finished():
+   req = state.CreateRequest()
+   resp = fetch([leader]/collect, req)
+   state.Update(resp)
+return state.Output()
+~~~
+
+[OPEN ISSUE: Describe how intra-protocol errors yield collect errors (see
+issue#57). For example, how does a leader respond to a collect request if the
+helper drops out?]
 
-## Verify {#pa-verify}
+Each collect request involves one of the helpers specified by the PA parameters.
+If more than one helper is specified, the collector may issue the requests in
+any order.
 
-[TODO: Add an illustration of this sub-protocol.]
+### Verifying and Aggregating Reports {#pa-aggregate}
 
 After the client uploads a report to the leader, the leader and helper verify in
-zero knowledge that the proof is well-formed. The exact procedure for doing so
-is protocol specific, but all protocols have the same basic structure. In
-particular, the protocol is comprised of a sequence of HTTPS requests from the
-leader to the helper. At the end of this phase, the leader and helper will have
-decided whether a set of client inputs are valid. For each valid input, they
-proceed as described in {{pa-collect}}.
-
-The leader begins by collecting a sequence of reports that are all associated
-with the same PA task, helper URL, and helper HPKE config id. Let `[helper]`
-denote the the URL. The leader sends a POST request to `[helper]/verify` with
-the following message:
+zero knowledge that the input is valid. The exact procedure for doing so is
+protocol specific, but all protocols have the same basic structure. In
+particular, the protocol is comprised of a sequence of *aggregate requests* from
+the leader to the helper. At the end of this procedure, the leader and helper
+will have have aggregated a set of valid client inputs.
+
+#### Aggregate Request
+
+The process begins with a PACollectReq. The leader collects a sequence of
+reports that are all associated with the same PA task, helper URL, and helper
+HPKE config id. Let `[helper]` denote `PACollectReq.helper_url`. The leader
+sends a POST request to `[helper]/aggregate` with the following message:
 
 ~~~
 struct {
   PATask task;
-  uint8 hpke_config_id;
-  PAVerifyReq seq<1..2^24-1>;
-} PAVerifyReqSeq;
+  uint8 helper_hpke_config_id;
+  opaque helper_state<0..2^16>;
+  PAAggregateSubReq seq<1..2^24-1>;
+} PAAggregateReq;
 ~~~
 
-The structure contains the PA task, the HPKE config id, and a sequence of
-*sub-requests*, each corresponding to a unique client report. Sub-requests are
-structured as follows:
+The structure contains the PA task, the helper's HPKE config id, an opaque
+*helper state* string, and a sequence of *sub-requests*, each corresponding to a
+unique client report.  Sub-requests are structured as follows:
 
 ~~~
 struct {
-  opaque enc<1..2^16-1>;
+  opaque enc<0..2^16-1>;
+  uint64 time; // UNIX timestamp (seconds) of the report.
```

Comment:
This is generated by the leader, presumably? Is it set when the report is received from the client, or when aggregation is actually done during collection? Could the client set the timestamp in its encrypted share for the helper?

Reply:
What stops the leader from altering the timestamp, then? Should it be folded into the helper AAD, perhaps?

Reply:
Yup, that's the idea. There's a TODO for this in "Upload Finish Request".